developer_manual/digging_deeper/translation.rst

```diff
@@ -0,0 +1,144 @@
+        public function getName(): string {
+            return $this->l->t('My awesome translation provider');
+        }
+
+        public function getAvailableLanguages(): array {
+            // Return an array of OCP\Translation\LanguageTuple objects here
+        }
+
+        public function translate(?string $fromLanguage, string $toLanguage, string $text): string {
+            // Do some fancy machine translation and return translated string
+        }
+    }
+
+The method ``getName`` returns a string to identify the registered provider in the user interface.
+
+The method ``translate`` translates the passed string and returns the translation. The two language parameters will be language codes that were returned by ``getAvailableLanguages`` of your provider. In case translation fails, you should throw a ``RuntimeException`` with an explanatory error message.
+
+The class would typically be saved into a file in ``lib/Translation`` of your app but you are free to put it elsewhere as long as it's loadable by Nextcloud's :ref:`dependency injection container<dependency-injection>`.
+
+Providing language detection
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There is also an ``IDetectLanguageProvider`` interface that allows specifying that your provider can auto-detect languages from text input. You can use this as follows:
+
+.. code-block:: php
+
+    <?php
+
+    declare(strict_types=1);
+
+    namespace OCA\MyApp\Translation;
+
+    use OCA\MyApp\AppInfo\Application;
+    use OCP\Files\File;
+    use OCP\Translation\ITranslationProvider;
+    use OCP\Translation\IDetectLanguageProvider;
+    use OCP\IL10N;
+
+    class Provider implements ITranslationProvider, IDetectLanguageProvider {
+
+        public function __construct(
+            private IL10N $l,
+        ) {
+        }
+
+        public function getName(): string {
+            return $this->l->t('My awesome translation provider');
+        }
+
+        public function getAvailableLanguages(): array {
+            // Return an array of OCP\Translation\LanguageTuple objects here
+        }
+
+        public function translate(?string $fromLanguage, string $toLanguage, string $text): string {
+            // Do some fancy machine translation and return translated string
+        }
+
+        public function detectLanguage(string $text): ?string {
+            // Detect the language of $text
+        }
+    }
+
+For convenience, when your provider implements language detection, ``null`` will be passed as the first parameter to ``translate``, so you can decide how to most efficiently detect the language as part of the translation process.
+
+If your provider does not implement ``IDetectLanguageProvider``, ``$fromLanguage`` will never be null.
+
+The method ``detectLanguage`` takes a text in some language and outputs the code of that language, or ``null`` in case detection wasn't successful. The language code that this method returns should match the ones you return in ``getAvailableLanguages``.
```

Comment:
```suggestion
The method ``detectLanguage`` takes a text in some language and outputs the code of that language, or ``null`` in case detection wasn't successful. The language code that this method returns should be one of the languages returned in ``getAvailableLanguages``.

should or has to be?
```